worker/uniter/runner/context/context_test.go

```diff
@@ -644,6 +644,48 @@ func (s *mockHookContextSuite) TestSetCacheStateErr(c *gc.C) {
 	c.Assert(err, gc.ErrorMatches, "loading unit state from database: testing an error")
 }
 
+func (s *mockHookContextSuite) TestFlushWithNonDirtyCache(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	hookContext := context.NewMockUnitHookContext(s.mockUnit)
+
+	s.mockUnit.EXPECT().State().Return(map[string]string{
+		"foo": "bar",
+	}, nil)
+
+	// The following commands are no-ops as they don't mutate the cache.
+	hookContext.SetCacheValue("foo", "bar")   // no-op: KV already present
+	hookContext.DeleteCacheValue("not-there") // no-op: key not present
+
+	// Flush the context with a success. As the cache is not dirty we do
+	// not expect a SetState call.
+	err := hookContext.Flush("success", nil)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *mockHookContextSuite) TestSequentialFlushOfCacheValues(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	hookContext := context.NewMockUnitHookContext(s.mockUnit)
+
+	// We expect a single call for the following API endpoints
+	s.mockUnit.EXPECT().State().Return(map[string]string{
```

Comment:
you could use s.expectStateValues() here as a replacement.

Reply:
Good idea. Let me add that in!